File: docs/chapter2/transformer-python/transformer_pytorch/core/transformer.py
"""
完整的 Transformer 模型

实现端到端的 Transformer 架构，包括：
- 完整的编码器-解码器 Transformer
- 仅编码器模型（如 BERT）
- 仅解码器模型（如 GPT）
- 序列到序列任务支持
- 语言建模任务支持

作者: Transformer-PyTorch Team
版本: 1.0.0
"""

from typing import Dict, Optional, Tuple, Union

import torch
import torch.nn as nn
import torch.nn.functional as F
from torch import Tensor

from .attention import AttentionMask
from .decoder import TransformerDecoder, create_decoder
from .embedding import TransformerEmbedding
from .encoder import TransformerEncoder, create_encoder
from .layers import LayerNorm
from .math_utils import init_weights
from ..config.config import TransformerConfig


class Transformer(nn.Module):
    """
    完整的 Transformer 模型
    
    实现编码器-解码器架构的 Transformer，支持序列到序列任务。
    
    Args:
        config (TransformerConfig): 模型配置
        
    Examples:
        >>> config = TransformerConfig(vocab_size=10000, d_model=512)
        >>> model = Transformer(config)
        >>> src = torch.randint(0, 10000, (2, 10))
        >>> tgt = torch.randint(0, 10000, (2, 8))
        >>> output = model(src, tgt)
        >>> print(output.logits.shape)  # torch.Size([2, 8, 10000])
    """
    
    def __init__(self, config: TransformerConfig):
        super().__init__()
        self.config = config
        
        # 源序列嵌入层
        self.src_embedding = TransformerEmbedding(
            vocab_size=config.vocab_size,
            d_model=config.d_model,
            max_len=config.max_seq_len,
            padding_idx=config.pad_token_id,
            position_encoding_type=config.position_encoding_type,
            dropout=config.dropout,
            scale_embedding=config.scale_embedding
        )
        
        # 目标序列嵌入层
        if config.tie_word_embeddings:
            self.tgt_embedding = self.src_embedding
        else:
            self.tgt_embedding = TransformerEmbedding(
                vocab_size=config.vocab_size,
                d_model=config.d_model,
                max_len=config.max_seq_len,
                padding_idx=config.pad_token_id,
                position_encoding_type=config.position_encoding_type,
                dropout=config.dropout,
                scale_embedding=config.scale_embedding
            )
        
        # 编码器
        if config.num_encoder_layers > 0:
            self.encoder = create_encoder(
                d_model=config.d_model,
                num_heads=config.num_heads,
                d_ff=config.d_ff,
                num_layers=config.num_encoder_layers,
                dropout=config.dropout,
                activation=config.activation,
                norm_first=config.norm_first
            )
        else:
            self.encoder = None
        
        # 解码器
        if config.num_decoder_layers > 0:
            self.decoder = create_decoder(
                d_model=config.d_model,
                num_heads=config.num_heads,
                d_ff=config.d_ff,
                num_layers=config.num_decoder_layers,
                dropout=config.dropout,
                activation=config.activation,
                norm_first=config.norm_first
            )
        else:
            self.decoder = None
        
        # 输出投影层
        self.output_projection = nn.Linear(config.d_model, config.vocab_size, bias=False)
        
        # 权重共享
        if config.tie_word_embeddings:
            self.output_projection.weight = self.src_embedding.token_embedding.embedding.weight
        
        # 初始化权重
        self.apply(lambda module: init_weights(module, config.init_std))
    
    def forward(
        self,
        src: Optional[Tensor] = None,
        tgt: Optional[Tensor] = None,
        src_mask: Optional[Tensor] = None,
        tgt_mask: Optional[Tensor] = None,
        memory_mask: Optional[Tensor] = None,
        return_dict: bool = True,
        **kwargs
    ) -> Union[Tensor, Dict[str, Tensor]]:
        """
        前向传播
        
        Args:
            src (Optional[Tensor]): 源序列 [batch_size, src_len]
            tgt (Optional[Tensor]): 目标序列 [batch_size, tgt_len]
            src_mask (Optional[Tensor]): 源序列掩码
            tgt_mask (Optional[Tensor]): 目标序列掩码
            memory_mask (Optional[Tensor]): 记忆掩码
            return_dict (bool): 是否返回字典格式的输出
            
        Returns:
            Union[Tensor, Dict[str, Tensor]]: 模型输出
        """
        # 编码阶段
        if src is not None and self.encoder is not None:
            # 源序列嵌入
            src_embedded = self.src_embedding(src)
            
            # 创建源序列填充掩码
            if src_mask is None:
                src_mask = self._create_padding_mask(src, self.config.pad_token_id)
            
            # 编码器前向传播
            encoder_output = self.encoder(src_embedded, mask=src_mask)
            
            if isinstance(encoder_output, dict):
                memory = encoder_output['last_hidden_state']
                encoder_attentions = encoder_output.get('all_attention_weights', None)
            else:
                memory = encoder_output
                encoder_attentions = None
        else:
            memory = None
            encoder_attentions = None
        
        # 解码阶段
        if tgt is not None and self.decoder is not None:
            # 目标序列嵌入
            tgt_embedded = self.tgt_embedding(tgt)
            
            # 创建目标序列因果掩码
            if tgt_mask is None:
                tgt_len = tgt.size(1)
                tgt_mask = AttentionMask.create_causal_mask(tgt_len, device=tgt.device)
                tgt_mask = tgt_mask.unsqueeze(0).expand(tgt.size(0), -1, -1)
            
            # 创建记忆掩码（如果有编码器输出）
            if memory is not None and memory_mask is None:
                memory_mask = src_mask
            
            # 解码器前向传播
            decoder_output = self.decoder(
                tgt_embedded, memory,
                tgt_mask=tgt_mask,
                memory_mask=memory_mask
            )
            
            if isinstance(decoder_output, dict):
                hidden_states = decoder_output['last_hidden_state']
                decoder_self_attentions = decoder_output.get('all_self_attention_weights', None)
                decoder_cross_attentions = decoder_output.get('all_cross_attention_weights', None)
            else:
                hidden_states = decoder_output
                decoder_self_attentions = None
                decoder_cross_attentions = None
        elif memory is not None:
            # 仅编码器模式
            hidden_states = memory
            decoder_self_attentions = None
            decoder_cross_attentions = None
        else:
            raise ValueError("必须提供 src 或 tgt 中的至少一个")
        
        # 输出投影
        logits = self.output_projection(hidden_states)
        
        if return_dict:
            return {
                'logits': logits,
                'last_hidden_state': hidden_states,
                'encoder_last_hidden_state': memory,
                'encoder_attentions': encoder_attentions,
                'decoder_self_attentions': decoder_self_attentions,
                'decoder_cross_attentions': decoder_cross_attentions,
            }
        else:
            return logits
    
    def encode(
        self,
        src: Tensor,
        src_mask: Optional[Tensor] = None
    ) -> Dict[str, Tensor]:
        """
        仅编码
        
        Args:
            src (Tensor): 源序列
            src_mask (Optional[Tensor]): 源序列掩码
            
        Returns:
            Dict[str, Tensor]: 编码器输出
        """
        if self.encoder is None:
            raise ValueError("模型没有编码器")
        
        src_embedded = self.src_embedding(src)
        
        if src_mask is None:
            src_mask = self._create_padding_mask(src, self.config.pad_token_id)
        
        encoder_output = self.encoder(src_embedded, mask=src_mask)
        
        if isinstance(encoder_output, dict):
            return encoder_output
        else:
            return {'last_hidden_state': encoder_output}
    
    def decode(
        self,
        tgt: Tensor,
        memory: Tensor,
        tgt_mask: Optional[Tensor] = None,
        memory_mask: Optional[Tensor] = None
    ) -> Dict[str, Tensor]:
        """
        仅解码
        
        Args:
            tgt (Tensor): 目标序列
            memory (Tensor): 编码器输出
            tgt_mask (Optional[Tensor]): 目标序列掩码
            memory_mask (Optional[Tensor]): 记忆掩码
            
        Returns:
            Dict[str, Tensor]: 解码器输出
        """
        if self.decoder is None:
            raise ValueError("模型没有解码器")
        
        tgt_embedded = self.tgt_embedding(tgt)
        
        if tgt_mask is None:
            tgt_len = tgt.size(1)
            tgt_mask = AttentionMask.create_causal_mask(tgt_len, device=tgt.device)
            tgt_mask = tgt_mask.unsqueeze(0).expand(tgt.size(0), -1, -1)
        
        decoder_output = self.decoder(
            tgt_embedded, memory,
            tgt_mask=tgt_mask,
            memory_mask=memory_mask
        )
        
        if isinstance(decoder_output, dict):
            hidden_states = decoder_output['last_hidden_state']
        else:
            hidden_states = decoder_output
        
        logits = self.output_projection(hidden_states)
        
        return {
            'logits': logits,
            'last_hidden_state': hidden_states,
        }
    
    def generate(
        self,
        src: Optional[Tensor] = None,
        tgt: Optional[Tensor] = None,
        max_length: int = 50,
        num_beams: int = 1,
        temperature: float = 1.0,
        top_k: int = 50,
        top_p: float = 1.0,
        do_sample: bool = False,
        **kwargs
    ) -> Tensor:
        """
        生成序列
        
        Args:
            src (Optional[Tensor]): 源序列（用于序列到序列任务）
            tgt (Optional[Tensor]): 目标序列前缀
            max_length (int): 最大生成长度
            num_beams (int): 束搜索大小
            temperature (float): 温度参数
            top_k (int): Top-K 采样
            top_p (float): Top-P 采样
            do_sample (bool): 是否使用采样
            
        Returns:
            Tensor: 生成的序列
        """
        self.eval()
        
        with torch.no_grad():
            # 编码阶段（如果有源序列）
            if src is not None:
                encoder_output = self.encode(src)
                memory = encoder_output['last_hidden_state']
            else:
                memory = None
            
            # 初始化目标序列
            if tgt is None:
                batch_size = src.size(0) if src is not None else 1
                tgt = torch.full(
                    (batch_size, 1),
                    self.config.bos_token_id,
                    dtype=torch.long,
                    device=next(self.parameters()).device
                )
            
            # 简单的贪心解码
            for _ in range(max_length):
                # 解码当前序列
                if memory is not None:
                    decoder_output = self.decode(tgt, memory)
                else:
                    decoder_output = self.forward(tgt=tgt, return_dict=True)
                
                # 获取下一个词元的 logits
                next_token_logits = decoder_output['logits'][:, -1, :]
                
                # 应用温度
                if temperature != 1.0:
                    next_token_logits = next_token_logits / temperature
                
                # 选择下一个词元
                if do_sample:
                    # 采样
                    probs = F.softmax(next_token_logits, dim=-1)
                    next_token = torch.multinomial(probs, num_samples=1)
                else:
                    # 贪心选择
                    next_token = torch.argmax(next_token_logits, dim=-1, keepdim=True)
                
                # 添加到序列
                tgt = torch.cat([tgt, next_token], dim=1)
                
                # 检查是否生成了结束标记
                if (next_token == self.config.eos_token_id).all():
                    break
            
            return tgt
    
    def _create_padding_mask(self, token_ids: Tensor, pad_token_id: int) -> Tensor:
        """创建填充掩码"""
        return AttentionMask.create_padding_mask(token_ids, pad_token_id)
    
    def get_input_embeddings(self) -> nn.Module:
        """获取输入嵌入层"""
        return self.src_embedding.token_embedding.embedding
    
    def set_input_embeddings(self, embeddings: nn.Module) -> None:
        """设置输入嵌入层"""
        self.src_embedding.token_embedding.embedding = embeddings
        if not self.config.tie_word_embeddings:
            self.tgt_embedding.token_embedding.embedding = embeddings
    
    def get_output_embeddings(self) -> nn.Module:
        """获取输出嵌入层"""
        return self.output_projection
    
    def set_output_embeddings(self, embeddings: nn.Module) -> None:
        """设置输出嵌入层"""
        self.output_projection = embeddings


class TransformerForSequenceToSequence(Transformer):
    """
    用于序列到序列任务的 Transformer 模型
    """
    
    def __init__(self, config: TransformerConfig):
        if config.num_encoder_layers == 0 or config.num_decoder_layers == 0:
            raise ValueError("序列到序列模型需要编码器和解码器")
        super().__init__(config)


class TransformerForLanguageModeling(nn.Module):
    """
    用于语言建模任务的 Transformer 模型（仅解码器）
    """
    
    def __init__(self, config: TransformerConfig):
        super().__init__()
        self.config = config
        
        # 修改配置为仅解码器
        config.num_encoder_layers = 0
        
        # 嵌入层
        self.embedding = TransformerEmbedding(
            vocab_size=config.vocab_size,
            d_model=config.d_model,
            max_len=config.max_seq_len,
            padding_idx=config.pad_token_id,
            position_encoding_type=config.position_encoding_type,
            dropout=config.dropout,
            scale_embedding=config.scale_embedding
        )
        
        # 解码器（作为主干网络）
        self.transformer = create_decoder(
            d_model=config.d_model,
            num_heads=config.num_heads,
            d_ff=config.d_ff,
            num_layers=config.num_decoder_layers,
            dropout=config.dropout,
            activation=config.activation,
            norm_first=config.norm_first
        )
        
        # 语言建模头
        self.lm_head = nn.Linear(config.d_model, config.vocab_size, bias=False)
        
        # 权重共享
        if config.tie_word_embeddings:
            self.lm_head.weight = self.embedding.token_embedding.embedding.weight
        
        # 初始化权重
        self.apply(lambda module: init_weights(module, config.init_std))
    
    def forward(
        self,
        input_ids: Tensor,
        attention_mask: Optional[Tensor] = None,
        labels: Optional[Tensor] = None,
        return_dict: bool = True
    ) -> Union[Tensor, Dict[str, Tensor]]:
        """
        前向传播
        
        Args:
            input_ids (Tensor): 输入词元ID [batch_size, seq_len]
            attention_mask (Optional[Tensor]): 注意力掩码
            labels (Optional[Tensor]): 标签（用于计算损失）
            return_dict (bool): 是否返回字典格式
            
        Returns:
            Union[Tensor, Dict[str, Tensor]]: 模型输出
        """
        # 嵌入
        hidden_states = self.embedding(input_ids)
        
        # 创建因果掩码
        seq_len = input_ids.size(1)
        causal_mask = AttentionMask.create_causal_mask(seq_len, device=input_ids.device)
        causal_mask = causal_mask.unsqueeze(0).expand(input_ids.size(0), -1, -1)
        
        # 组合掩码
        if attention_mask is not None:
            # 将填充掩码与因果掩码结合
            attention_mask = attention_mask.unsqueeze(1) * attention_mask.unsqueeze(2)
            causal_mask = causal_mask * attention_mask
        
        # Transformer 前向传播（使用解码器作为主干）
        # 注意：这里我们传递 None 作为 memory，因为这是仅解码器模型
        transformer_output = self.transformer(
            hidden_states, memory=None,
            tgt_mask=causal_mask, memory_mask=None
        )
        
        if isinstance(transformer_output, dict):
            hidden_states = transformer_output['last_hidden_state']
        else:
            hidden_states = transformer_output
        
        # 语言建模头
        logits = self.lm_head(hidden_states)
        
        loss = None
        if labels is not None:
            # 计算交叉熵损失
            shift_logits = logits[..., :-1, :].contiguous()
            shift_labels = labels[..., 1:].contiguous()
            loss = F.cross_entropy(
                shift_logits.view(-1, shift_logits.size(-1)),
                shift_labels.view(-1),
                ignore_index=self.config.pad_token_id
            )
        
        if return_dict:
            return {
                'logits': logits,
                'loss': loss,
                'last_hidden_state': hidden_states,
            }
        else:
            return logits


# 导出的类
__all__ = [
    'Transformer',
    'TransformerForSequenceToSequence',
    'TransformerForLanguageModeling',
]

File: learning/transformer-js/core/transformer.js
/**
 * 完整的 Transformer 模型
 * 组合编码器、解码器、嵌入层等组件构建完整的 Transformer 架构
 * 
 * @author shihom_wu
 * @version 1.0.0
 */

const { TransformerEmbedding } = require('./embedding');
const { TransformerEncoder, createEncoder } = require('./encoder');
const { TransformerDecoder, createDecoder } = require('./decoder');
const { Linear } = require('./layers');
const { MaskGenerator } = require('./attention');
const { softmax } = require('./math-utils');
const { matmul } = require('./matrix-ops');

/**
 * 完整的 Transformer 模型
 * 实现编码器-解码器架构的 Transformer
 */
class Transformer {
  /**
   * 构造函数
   * 
   * @param {Object} config - 模型配置
   */
  constructor(config) {
    this.config = config;
    this.vocabSize = config.vocab_size;
    this.dModel = config.n_embd || config.d_model;
    this.maxSeqLen = config.max_seq_len;
    this.nLayers = config.n_layers;
    this.nHeads = config.n_heads;
    
    // 源序列嵌入层
    this.srcEmbedding = new TransformerEmbedding(
      this.vocabSize,
      this.dModel,
      this.maxSeqLen,
      config.position_encoding_type || 'sinusoidal',
      config.dropout || 0.1
    );
    
    // 目标序列嵌入层
    this.tgtEmbedding = new TransformerEmbedding(
      this.vocabSize,
      this.dModel,
      this.maxSeqLen,
      config.position_encoding_type || 'sinusoidal',
      config.dropout || 0.1
    );
    
    // 编码器
    this.encoder = createEncoder(config);
    
    // 解码器
    this.decoder = createDecoder(config);
    
    // 输出投影层（语言模型头）
    this.outputProjection = new Linear(
      this.dModel,
      this.vocabSize,
      config.use_bias || false
    );
    
    // 是否共享嵌入权重
    this.tieWeights = config.tie_word_embeddings || false;
    if (this.tieWeights) {
      this._tieEmbeddingWeights();
    }
    
    // 训练模式标志
    this.training = true;
  }
  
  /**
   * 共享输入输出嵌入权重
   * 将输出投影层的权重与词嵌入权重绑定
   */
  _tieEmbeddingWeights() {
    // 简化实现：在实际应用中需要确保权重矩阵的引用一致性
    console.log('权重共享已启用（简化实现）');
  }
  
  /**
   * 前向传播
   * 
   * @param {Array<Array<number>>} srcTokens - 源序列词元 [batchSize, srcSeqLen]
   * @param {Array<Array<number>>} tgtTokens - 目标序列词元 [batchSize, tgtSeqLen]
   * @param {Array<Array<number>>|null} srcMask - 源序列掩码，可选
   * @param {Array<Array<number>>|null} tgtMask - 目标序列掩码，可选
   * @returns {Object} 模型输出
   */
  forward(srcTokens, tgtTokens, srcMask = null, tgtMask = null) {
    // 验证输入
    if (!Array.isArray(srcTokens) || !Array.isArray(tgtTokens)) {
      throw new Error('源序列和目标序列必须是二维数组');
    }
    
    const batchSize = srcTokens.length;
    const srcSeqLen = srcTokens[0].length;
    const tgtSeqLen = tgtTokens[0].length;
    
    // 处理单个样本的情况（简化实现）
    if (batchSize === 1) {
      return this._forwardSingle(srcTokens[0], tgtTokens[0], srcMask, tgtMask);
    }
    
    // 批处理实现（简化）
    const results = [];
    for (let b = 0; b < batchSize; b++) {
      const result = this._forwardSingle(srcTokens[b], tgtTokens[b], srcMask, tgtMask);
      results.push(result);
    }
    
    return this._combineBatchResults(results);
  }
  
  /**
   * 单个样本的前向传播
   * 
   * @param {Array<number>} srcTokens - 源序列词元 [srcSeqLen]
   * @param {Array<number>} tgtTokens - 目标序列词元 [tgtSeqLen]
   * @param {Array<Array<number>>|null} srcMask - 源序列掩码
   * @param {Array<Array<number>>|null} tgtMask - 目标序列掩码
   * @returns {Object} 单个样本的输出
   */
  _forwardSingle(srcTokens, tgtTokens, srcMask = null, tgtMask = null) {
    // 源序列嵌入
    const srcEmbedded = this.srcEmbedding.forward([srcTokens])[0]; // [srcSeqLen, dModel]
    
    // 目标序列嵌入
    const tgtEmbedded = this.tgtEmbedding.forward([tgtTokens])[0]; // [tgtSeqLen, dModel]
    
    // 编码器前向传播
    const encoderResult = this.encoder.forward(srcEmbedded, srcMask);
    const encoderOutput = encoderResult.output; // [srcSeqLen, dModel]
    
    // 生成目标序列的因果掩码
    if (tgtMask === null) {
      tgtMask = MaskGenerator.createCausalMask(tgtTokens.length);
    }
    
    // 解码器前向传播
    const decoderResult = this.decoder.forward(
      tgtEmbedded,
      encoderOutput,
      tgtMask,
      srcMask
    );
    const decoderOutput = decoderResult.output; // [tgtSeqLen, dModel]
    
    // 输出投影到词汇表
    const logits = this.outputProjection.forward(decoderOutput); // [tgtSeqLen, vocabSize]
    
    return {
      logits: logits,
      encoderOutput: encoderOutput,
      encoderAttentions: encoderResult.attentions,
      decoderSelfAttentions: decoderResult.selfAttentions,
      decoderCrossAttentions: decoderResult.crossAttentions
    };
  }
  
  /**
   * 组合批处理结果
   * 
   * @param {Array<Object>} results - 各样本的结果
   * @returns {Object} 组合后的批处理结果
   */
  _combineBatchResults(results) {
    return {
      logits: results.map(r => r.logits),
      encoderOutputs: results.map(r => r.encoderOutput),
      encoderAttentions: results.map(r => r.encoderAttentions),
      decoderSelfAttentions: results.map(r => r.decoderSelfAttentions),
      decoderCrossAttentions: results.map(r => r.decoderCrossAttentions)
    };
  }
  
  /**
   * 编码
   * 只运行编码器部分
   * 
   * @param {Array<Array<number>>} srcTokens - 源序列词元
   * @param {Array<Array<number>>|null} srcMask - 源序列掩码
   * @returns {Object} 编码器输出
   */
  encode(srcTokens, srcMask = null) {
    const batchSize = srcTokens.length;
    
    if (batchSize === 1) {
      const srcEmbedded = this.srcEmbedding.forward(srcTokens)[0];
      return this.encoder.forward(srcEmbedded, srcMask);
    }
    
    // 批处理编码
    const results = [];
    for (let b = 0; b < batchSize; b++) {
      const srcEmbedded = this.srcEmbedding.forward([srcTokens[b]])[0];
      const result = this.encoder.forward(srcEmbedded, srcMask);
      results.push(result);
    }
    
    return {
      outputs: results.map(r => r.output),
      attentions: results.map(r => r.attentions)
    };
  }
  
  /**
   * 解码
   * 给定编码器输出，运行解码器
   * 
   * @param {Array<Array<number>>} tgtTokens - 目标序列词元
   * @param {Array<Array<number>>} encoderOutput - 编码器输出
   * @param {Array<Array<number>>|null} tgtMask - 目标序列掩码
   * @param {Array<Array<number>>|null} srcMask - 源序列掩码
   * @returns {Object} 解码器输出
   */
  decode(tgtTokens, encoderOutput, tgtMask = null, srcMask = null) {
    const batchSize = tgtTokens.length;
    
    if (batchSize === 1) {
      const tgtEmbedded = this.tgtEmbedding.forward(tgtTokens)[0];
      
      if (tgtMask === null) {
        tgtMask = MaskGenerator.createCausalMask(tgtTokens[0].length);
      }
      
      const decoderResult = this.decoder.forward(
        tgtEmbedded,
        encoderOutput,
        tgtMask,
        srcMask
      );
      
      const logits = this.outputProjection.forward(decoderResult.output);
      
      return {
        logits: logits,
        selfAttentions: decoderResult.selfAttentions,
        crossAttentions: decoderResult.crossAttentions
      };
    }
    
    // 批处理解码（简化实现）
    throw new Error('批处理解码暂未实现');
  }
  
  /**
   * 生成下一个词元的概率分布
   * 
   * @param {Array<Array<number>>} srcTokens - 源序列
   * @param {Array<number>} tgtTokens - 目标序列（到当前位置）
   * @returns {Array<number>} 下一个词元的概率分布
   */
  predictNext(srcTokens, tgtTokens) {
    // 编码源序列
    const encoderResult = this.encode([srcTokens]);
    const encoderOutput = encoderResult.outputs[0];
    
    // 解码到当前位置
    const decoderResult = this.decode([tgtTokens], encoderOutput);
    const logits = decoderResult.logits;
    
    // 获取最后一个位置的 logits
    const lastLogits = logits[logits.length - 1];
    
    // 转换为概率分布
    return softmax(lastLogits);
  }
  
  /**
   * 设置训练模式
   * 
   * @param {boolean} training - 是否为训练模式
   */
  setTraining(training) {
    this.training = training;
    this.srcEmbedding.setTraining(training);
    this.tgtEmbedding.setTraining(training);
    this.encoder.setTraining(training);
    this.decoder.setTraining(training);
  }
  
  /**
   * 获取模型参数数量
   * 
   * @returns {Object} 参数统计信息
   */
  getParameterCount() {
    const srcEmbeddingParams = this.srcEmbedding.getParameterCount();
    const tgtEmbeddingParams = this.tieWeights ? 0 : this.tgtEmbedding.getParameterCount();
    const encoderParams = this.encoder.getParameterCount();
    const decoderParams = this.decoder.getParameterCount();
    const outputParams = this.outputProjection.getParameterCount();
    
    const totalParams = srcEmbeddingParams + tgtEmbeddingParams + encoderParams + decoderParams + outputParams;
    
    return {
      srcEmbedding: srcEmbeddingParams,
      tgtEmbedding: tgtEmbeddingParams,
      encoder: encoderParams,
      decoder: decoderParams,
      outputProjection: outputParams,
      total: totalParams,
      totalM: (totalParams / 1e6).toFixed(2) + 'M'
    };
  }
  
  /**
   * 模型信息摘要
   * 
   * @returns {Object} 模型信息
   */
  summary() {
    const paramCount = this.getParameterCount();
    
    return {
      architecture: 'Transformer (Encoder-Decoder)',
      vocabSize: this.vocabSize,
      modelDim: this.dModel,
      maxSeqLen: this.maxSeqLen,
      layers: this.nLayers,
      heads: this.nHeads,
      parameters: paramCount,
      config: this.config
    };
  }
}

/**
 * 创建 Transformer 模型的工厂函数
 * 
 * @param {Object} config - 模型配置
 * @returns {Transformer} Transformer 模型实例
 */
function createTransformer(config) {
  // 验证配置
  const requiredParams = ['vocab_size', 'n_embd', 'n_layers', 'n_heads'];
  for (const param of requiredParams) {
    if (!(param in config)) {
      throw new Error(`缺少必需的配置参数: ${param}`);
    }
  }
  
  return new Transformer(config);
}

// 导出类和函数
module.exports = {
  Transformer,
  createTransformer
};
